### PR TITLE
feature for log in process for existing user complete

### PR DESCRIPTION
**Pull request by AndyPark20**
### 1. Feature complete - log in for existing user
---------------------------------
-  User can input y or n (yes/no) in the terminal to determine if the user is an existing user or new user:
image

- Once user enters Y (It will be down cased when checking for credentials in the code), It will prompt the user for last name, password, and 4 digits for verification. **validate_log_in_credentials** method will loop over @@user_info array of objects to check for objects that has the corresponding values present in the key/value pair. 

- **If found, it will welcome the user.** 
image

- **If credentials are wrong, it will deny the user**
image

### CODE REVIEW

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Generally best practice to avoid class variables
https://github.com/rubocop/ruby-style-guide#no-class-vars
Btw, feel free to reference the style guide in general! It's what we use!

Prefer to use constants. For example: 
```
USER_INFO = [
```
P.s. Make sure there are spaces around assignment operators.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Prefer spaces like so for readability:
```
{ first: 'Jason', last: 'kim', password: 'bye123', ssn: '135791098' }
```

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Prefer to assign instance variables within `initialize` or some other method instead of free-floating.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: 💯

Comment (gabe-kent) [reply bank_component.rb]: If these attributes don't need to be edited after the class is initialized, then it's best to restrict access to only what is needed.

In other words, `attr_reader` would make these uneditable from outside of the class.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Nit: spaces around the assignment operator

("Nit" as in "nitpick". It's part of the culture to label our nitpicks as such, so the author doesn't feel bashed. As in, "we recognize this is not a critical thing to fix, but we still want to give you a friendly reminder")

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Prefer to use descriptive param names in blocks. For example:
```
@@user_info.each do |user|
```

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: I assume this is to grab the last 4 digits. 
Another way to do this would be: 
```
x[:ssn][-4..-1]
```
This at least guarantees that the last 4 digits are always retrieved.

Another option to consider is using Regexes, but unless you're wanting to validate the format, it's probably unnecessary.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Ah, I see that this method is supposed to return once a match is found. 
There is a handy ruby method that could help shorten this whole method a bit and avoid having to explicitly `return`.

Check out Enumerable#find
https://apidock.com/ruby/Enumerable/find

P.S. Avoiding explicit returns is not a style requirement, but it generally means that things could be cleaner.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: This `else` block might have unintentional side effects. i.e. There might be a bug here.

I believe this `return` will cause the whole method to return if the first user does not match. Meaning, the second user will never be checked.

To fix this, we'll probably need to avoid these returns altogether. Though it would be fine to still return for the match case.

For the non-match case, we might need to remove the `else` completely, then output `"sorry access denied!"` after the `each` loop finishes. This way we will only make it past the each loop if all users failed to match the credentials. If there was a match, the early return would have skipped the message after the loop.

Another approach would be to use local vars to track the state of things.

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: 😆

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Nit: spaces
```
(user_last_name_input, user_password_input, ...
```

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: An if-statement would probably be more appropriate here unless you plan to add more cases

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: Global vars are extremely rare to use and can be very problematic. Avoid at all costs! 🔴 
Class Variables are likewise very rare and worth avoiding. 🟠 
Instance Variables and class Constants are good 🟢

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb:87]: I like the recursive nature of this!

Small nit: Maybe there should be a quit case to quickly exit. Like `when 'q'`

File: bank_component.rb
Comment (gabe-kent) [inline bank_component.rb]: It's best practice to avoid block comments
https://github.com/rubocop/ruby-style-guide#rd-block-comments

**Review by gabe-kent — CHANGES_REQUESTED** [review]
@AndyPark20 You got a solid script! Great job using OO principles here. I left some suggestions and ideas to consider. Don't feel bad about the nitpicks! It's all part of the initiation!

File: bank_component.rb
Comment (AndyPark20) [reply bank_component.rb]: Well noted!!

**Comment by gabe-kent** [comment]
@AndyPark20 Nice job on the changes! 

Normally after making changes the author would re-request a review from the code reviewer in order to get an explicit approval before merging. No worries though! We can incorporate that in the next PR. 👍

**Comment by AndyPark20** [comment]
> @AndyPark20 Nice job on the changes!
> 
> Normally after making changes the author would re-request a review from the code reviewer in order to get an explicit approval before merging. No worries though! We can incorporate that in the next PR. 👍

@gabe-kent well noted!
I will make sure to follow the correct procedure on my next pull request!

**Pull request merged**
